docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,456 @@
+
+```haskell
+type LedgerState :: Type -> MapKind -> Type
+data family LedgerState blk mk
+
+type LedgerTables :: (Type -> MapKind -> Type) -> MapKind -> Type
+newtype LedgerTables l mk = LedgerTables {
+    getLedgerTables :: mk (Key l) (Value l)
+  }
+```
+
+### Keeping the ledger states in memory (the ~LedgerDB~ DbChangelog)
+
+If we were to store only the resulting ledger state after applying a block, we
+would lose track of what values were consumed. On the other hand, if we kept a
+UTxO set next to each ledger state, we would just defeat the purpose of UTxO-HD.
+To solve this, we keep one copy of the UTxO set at some immutable block (usually
+the immutable tip, but not necessarily) and then the differences of each
+volatile block. Moving that copy of the UTxO set to disk is what achieves the
+memory decrease that we were seeking with this project.
+
+For efficiency reasons, instead of keeping a sequence of states carrying a
+`DiffMK` each, we keep two sequences: one of `EmptyMK` ledger states and one
+with all the `DiffMK`s grouped into a modified
+[`FingerTree`](https://github.com/input-output-hk/anti-diffs/tree/main/fingertree-rm).
+
+This need for a different data structure to hold the history of Ledger states
+prompted us to use a `DbChangelog` instead of the pre-UTxO-HD `LedgerDB`.
+Together with a pointer to the data in the disk, this is functionally equivalent
+to the old `LedgerDB` contents.
+
+```haskell
+data DbChangelog l = DbChangelog {
+    changelogLastFlushedState :: !(l EmptyMK)
+  , anchorlessChangelog       :: !(AnchorlessDbChangelog l)
+  }
+  
+data AnchorlessDbChangelog l = AnchorlessDbChangelog {
+    adcLastFlushedSlot :: !(WithOrigin SlotNo)
+  , adcDiffs           :: !(LedgerTables l SeqDiffMK)
+  , adcStates          :: !(StatesSequence l)
+  }
+  
+type StatesSequence l = 
+  AnchoredSeq (WithOrigin SlotNo) (l EmptyMK) (l EmptyMK)
+```
+
+```mermaid
+stateDiagram-v2
+    state "DbChangelog l" as dbch {
+        state "LastFlushedState :: l EmptyMK" as lfs
+        state StateSequence {
+            direction LR
+            state "ImmutableTip :: l EmptyMK" as tip
+            tip --> L1
+            L1 --> L2
+            state "..." as dots1
+            L2 --> dots1
+            dots1 --> Lk
+        } 
+        state Diffs {
+            direction LR
+            state "D-1" as dm1
+            state "D-2" as dm2
+            dm2 --> dm1
+            dm1 --> D0
+            D0 --> D1
+            D1 --> D2
+            state "..." as dots2
+            D2 --> dots2
+            dots2 --> Dk
+        } 
+    }
+```
+
+In this diagram, `D1` corresponds to the delta resulting between the states
+`ImmutableTip` and `L1`, `D2` is the delta between `L1` and `L2`, and so on.
+Notice the existence of deltas with negative numbers. These are the deltas
+belonging to the (already garbage-collected as they are not needed) immutable
+states between the `LastFlushedState` and the `ImmutableTip`.
```

Comment:
Maybe add a sentence explaining that we keep the "extra" diffs so that we can flush them?

Related question: would it be helpful to carry a single "extra" diff that is maintained as the simply composition of all the immutable-but-not-yet-flushed diffs?

Reply:
Added the sentence. For your other question, maybe? Maybe we could carry a seqdiff (fingertree) for the volatile chain but just a diff for the immutable chain? What do you think @jorisdral ?

Reply:
Referenced in #310